cpp/src/arrow/dataset/discovery.cc

```diff
@@ -100,42 +105,38 @@ Result<DataSourceDiscoveryPtr> FileSystemDataSourceDiscovery::Make(
               std::move(options));
 }
 
-static inline Result<std::shared_ptr<Schema>> InspectSchema(
-    fs::FileSystem* fs, const std::vector<fs::FileStats> stats,
-    const FileFormatPtr& format) {
+Result<std::shared_ptr<Schema>> FileSystemDataSourceDiscovery::Inspect() {
   std::vector<std::shared_ptr<Schema>> schemas;
 
-  for (const auto& f : stats) {
+  for (const auto& f : files_) {
     if (!f.IsFile()) continue;
 
-    ARROW_ASSIGN_OR_RAISE(auto schema, format->Inspect(FileSource(f.path(), fs)));
+    ARROW_ASSIGN_OR_RAISE(auto schema, format_->Inspect(FileSource(f.path(), fs_.get())));
     schemas.push_back(schema);
   }
 
-  if (schemas.size() > 0) {
-    // TODO merge schemas.
-    return schemas[0];
+  if (schemas.empty()) {
+    // If there are no files, return the partition scheme's schema.
+    return partition_scheme_->schema();
   }
 
-  // If there is no files, return an empty schema.
-  return std::shared_ptr<Schema>(nullptr);
-}
+  // TODO merge schemas.
+  auto out_schema = arrow::schema(schemas[0]->fields());
```

Comment:
This changes looses the metadata of the original schema (now, metadata is a topic that probably needs to be revisited anyhow when unifying / merging schema's)

Reply:
Whoops. For now I'll ensure that at least the metadata of schemas [0] is preserved